runtime/src/serde_snapshot.rs

```diff
@@ -326,19 +356,68 @@ where
     let mut accounts_db = AccountsDb::new_with_config(
         account_paths.to_vec(),
         cluster_type,
-        account_indexes,
+        account_secondary_indexes,
         caching_enabled,
         shrink_ratio,
     );
-    let AccountsDbFields(storage, version, slot, bank_hash_info) = accounts_db_fields;
+    let AccountsDbFields(
+        full_snapshot_storage,
+        full_snapshot_version,
+        full_snapshot_slot,
+        full_snapshot_bank_hash_info,
+    ) = snapshot_accounts_db_fields.full_snapshot_accounts_db_fields;
+
+    let (snapshot_version, snapshot_slot, snapshot_bank_hash_info) = snapshot_accounts_db_fields
+        .incremental_snapshot_accounts_db_fields
+        .as_ref()
+        .map(
+            |AccountsDbFields(
+                _,
+                incremental_snapshot_version,
+                incremental_snapshot_slot,
+                incremental_snapshot_bank_hash_info,
+            )| {
+                (
+                    *incremental_snapshot_version,
+                    *incremental_snapshot_slot,
+                    incremental_snapshot_bank_hash_info.clone(),
+                )
+            },
+        )
+        .unwrap_or((
+            full_snapshot_version,
+            full_snapshot_slot,
+            full_snapshot_bank_hash_info,
+        ));
+
+    // filter out incremental snapshot storages with slot <= full snapshot slot
+    let incremental_snapshot_storage = snapshot_accounts_db_fields
+        .incremental_snapshot_accounts_db_fields
+        .map_or_else(HashMap::new, |fields| {
+            fields
+                .0
+                .into_iter()
+                .filter(|(slot, _)| *slot > full_snapshot_slot)
+                .collect::<HashMap<_, _>>()
+        });
+
+    // There must not be any overlap in the slots of storages between the full snapshot and the incremental snapshot
+    incremental_snapshot_storage
+        .iter()
+        .all(|storage_entry| !full_snapshot_storage.contains_key(storage_entry.0)).then(|| ()).ok_or_else(|| {
+            io::Error::new(io::ErrorKind::InvalidData, "Snapshots are incompatible: There are storages for the same slot in both the full snapshot and the incremental snapshot!")
+        })?;
```

Comment:
how about extracting this code block to a newly-defined method at `SnapshotAccountsDbFields`?

Reply:
i think this incremental/full snapshot new logic can be well encapsulated, considering this following code (https://github.com/solana-labs/solana/pull/18504/files#r671230020) like this:

```rust

impl SnapshotAccountsDbFields {
    fn combined_storages(...) -> Vec<_> {
    }
}

```

Reply:
Done!